iothub/device/src/ClientPropertyCollection.cs

```diff
@@ -0,0 +1,378 @@
+        /// <param name="componentName">The component which holds the required property.</param>
+        /// <returns><c>true</c> if the specified property is present; otherwise, <c>false</c>.</returns>
+        public bool Contains(string componentName, string propertyName)
+        {
+            if (!string.IsNullOrEmpty(componentName) && Collection.TryGetValue(componentName, out var component))
+            {
+                return Convention.PayloadSerializer.TryGetNestedObjectValue<object>(component, propertyName, out _);
+            }
+            return Collection.TryGetValue(propertyName, out _);
+        }
+
+        /// <summary>
+        /// Gets the version of the property collection.
+        /// </summary>
+        /// <value>A <see cref="long"/> that is used to identify the version of the property collection.</value>
+        public long Version { get; protected set; }
+
+        /// <summary>
+        /// Gets the value of a component-level property.
+        /// </summary>
+        /// <remarks>
+        /// To get the value of a root-level property use <see cref="PayloadCollection.TryGetValue{T}(string, out T)"/>.
+        /// </remarks>
+        /// <typeparam name="T">The type to cast the object to.</typeparam>
+        /// <param name="componentName">The component which holds the required property.</param>
+        /// <param name="propertyName">The property to get.</param>
+        /// <param name="propertyValue">The value of the component-level property.</param>
+        /// <returns>true if the property collection contains a component level property with the specified key; otherwise, false.</returns>
+        public virtual bool TryGetValue<T>(string componentName, string propertyName, out T propertyValue)
+        {
+            if (Contains(componentName, propertyName))
+            {
+                object componentProperties = Collection[componentName];
+                Convention.PayloadSerializer.TryGetNestedObjectValue<T>(componentProperties, propertyName, out propertyValue);
+                return true;
+            }
+
+            propertyValue = default;
+            return false;
+        }
+
+        /// <summary>
+        /// Converts a <see cref="TwinCollection"/> collection to a properties collection.
+        /// </summary>
+        /// <remarks>This internal class is aware of the implementation of the TwinCollection ad will </remarks>
+        /// <param name="twinCollection">The TwinCollection object to convert.</param>
+        /// <param name="payloadConvention">A convention handler that defines the content encoding and serializer to use for the payload.</param>
+        /// <returns>A new instance of the class from an existing <see cref="TwinProperties"/> using an optional <see cref="PayloadConvention"/>.</returns>
+        internal static ClientPropertyCollection FromTwinCollection(TwinCollection twinCollection, PayloadConvention payloadConvention)
```

Comment:
the plan is to merge the `Twin` and `Property` flows, post preview and before GA.